### PR TITLE
Improving variable search mechanism

### PR DESCRIPTION
**Pull request by akuleshov7**
### What's done:
1) Adding mechanism for searching assignments of VARs
2) Moving all usages to new utilities of searching variables

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt [hunk 1]
```diff
@@ -160,7 +159,6 @@ fun ASTNode.findChildAfter(afterThisNodeType: IElementType, childNodeType: IElem
         }
     }
 
-    log.warn("Not able to find a node with type $childNodeType after $afterThisNodeType")
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt:163]: Does this mean we can close #141 in this PR?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/search/VariablesSearch.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                            ?: (this.getParentOfType<KtClassBody>(true)?.getAllSearchResults(this))
+                            // searching on the file level
+                            ?: (this.getParentOfType<KtFile>(true)!!.getAllSearchResults(this))
+                }
+    }
+
+    /**
+     * filtering object's fields (expressions) that have same name as variable
+     */
+    protected fun isReferenceToFieldOfObject(expression: KtNameReferenceExpression) =
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/search/VariablesSearch.kt:60]: Maybe make it an extension on `KtNameReferenceExpression`?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/search/VariablesWithAssignmentSearch.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.jetbrains.kotlin.psi.KtProperty
+
+class VariablesWithAssignmentSearch(fileNode: ASTNode,
+                                    filterForVariables: (KtProperty) -> Boolean) : VariableSearch(fileNode, filterForVariables) {
+
+    override fun KtElement.getAllSearchResults(property: KtProperty): List<KtNameReferenceExpression> {
+        return this.node.findAllNodesWithSpecificType(ElementType.BINARY_EXPRESSION)
+                // filtering out all usages that are declared in the same context but are going before the variable declaration
+                .filter {
+                    // FixMe: bug is here with a search of postfix/prefix variables assignment
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/search/VariablesWithAssignmentSearch.kt:19]: What do you mean? Operation like `++`, `+=`?

Comment (akuleshov7) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/search/VariablesWithAssignmentSearch.kt:19]: yes, they will not work properly now, I will handle it

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt (hunk 1 above)
Comment (akuleshov7) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt:163]: yes, I decided to remove it. You were right - it does not make any sense

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/search/VariablesSearch.kt (hunk 2 above)
Comment (akuleshov7) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/search/VariablesSearch.kt]: fixed

**Review by petertrr — APPROVED** [review]

**Pull request merged**